README.md
# aws_longer

`aws_longer` is a tool to replace
[assume-role](https://github.com/coinbase/assume-role) with a few significant
benefits:

1) MFA tokens need only be entered once every 36 hours.
2) Sessions can be shared across terminals, and persist across reboots.

Both of the above are accomplished by storing the session credentials, and
assumed credentials in the user's keychain.

## Installation

Install this package via:

```sh
pip install aws_longer
```

If you'd like to use this package with your yubikey please run:

```sh
pip install aws_longer[yubikey]  # Escape the brackets in zsh
```

## Assuming a Role

```sh
aws_longer role ROLENAME AWS_ACCOUNT
```

The above will open a new shell setting the appropriate `AWS_` environment
variables. If this is the first time you are running this command, or it has
been 36 hours since you last input your MFA token, then you will be prompted to
input your MFA token.

`AWS_ACCOUNT` can either be an AWS account ID, or an alias to an AWS account
ID.

If you'd like to prevent opening a new shell, you can instead run, but be
careful because if there are any errors, the result will terminate your
shell-program:

```sh
exec aws_longer role ROLENAME AWS_ACCOUNT
```

## Using the Temporary Session

Rather than assuming a specific role, one can directly utilize the 36-hour
temporary session via:

```sh
aws_longer
```

Using this temporary session is beneficial if, for example, you require MFA to
assume roles, and you'd like to be able to run `terraform apply` with a
provider that assumes a specific role.


## AWS Account ID Alias

AWS account ID aliases can be defined in `~/.aws/accounts`, which is a JSON
file of the following format:

```json
{
  "default": "123456789012",
  "staging": "123456789012",
  "production": "123456789012"
}
```

This aliasing format is the same as was used in
[assume-role](https://github.com/coinbase/assume-role#account-aliasing) in
order to ease transitioning.

## Avoid token prompting

If you happen to use a program to generate your MFA token, you might prefer to
pass the token directly as part of the command invocation instead of via a
prompt to STDIN. Do that via:

```sh
aws_longer --mfa-token TOKEN role ROLENAME AWS_ACCOUNT
```

In said cases, the MFA token will only be used if it's necessary to do so.

### Yubikeys

If you are using a Yubikey in TOTP mode, you can use it via:

```sh
aws_longer --yubikey NAME role ROLENAME AWS_ACCOUNT
```

To discover the name on your yubikey, you can list all names via:

```sh
ykman oath list
```


CHANGES.md
# Change Log

## 0.4.0 (2021/04/28)

__Changed__

* Use IAM `UserName` for `RoleSession` name instead of the OS username.

## 0.3.0 (2020/04/17)

__Added__

* Support yubikeys via the `--yubikey NAME` argument.

__Fixed__

* `--cleanup` works again for clearing the temporary credentials (non-role
  based one)

## 0.2.1 (2020/04/03)

__Fixed__

* Only prompt for session MFA when it actually is needed.
* Avoid pulling in a random role token as the session token when the session
  token expires.

## 0.2.0 (2020/03/06)

__Added__

* Support MFA input via `--mfa-token` command line option.

## 0.1.2 (2020/03/06)

__Fixed__

* Declare boto3 as a dependency

## 0.1.1 (2020/03/06)

__Fixed__

* Declare keyring as a dependency

## 0.1.0 (2020/03/06)

Initial version.


tests/test_aws_longer.py
from datetime import datetime, timedelta, timezone

import aws_longer
import keyring

aws_longer.KEYRING_SERVICE_NAME = "aws_longer_test"


def setup():
    for username in (aws_longer.SESSION_KEYRING_USERNAME, "account_role"):
        try:
            keyring.delete_password(
                service_name=aws_longer.KEYRING_SERVICE_NAME, username=username
            )
        except keyring.errors.PasswordDeleteError:
            pass


def test_cache_in_keyring__change_account_name():
    function = aws_longer.cache_in_keyring(lambda x, **kwargs: x)
    now = datetime.now(timezone.utc)
    expected = {"Expiration": now + timedelta(minutes=2)}

    assert expected == function(expected, account="account", role="role")
    assert expected == function({"Expiration": now}, account="account", role="role")


def test_cache_in_keyring__use_cached_value():
    function = aws_longer.cache_in_keyring(lambda x: x)
    now = datetime.now(timezone.utc)
    expected = {"Expiration": now + timedelta(minutes=2)}

    assert expected == function(expected)
    assert expected == function({"Expiration": now})


def test_cache_in_keyring__dont_use_expired_value():
    function = aws_longer.cache_in_keyring(lambda x: x)
    now = datetime.now(timezone.utc)
    expected = {"Expiration": now + timedelta(minutes=1)}

    assert expected == function(expected)
    assert expected != function({"Expiration": now})


aws_session.py
#!/usr/bin/env python3
import argparse
import base64
import json
import os
import pickle
import sys
from datetime import datetime, timedelta, timezone
from functools import wraps

import boto3
import botocore
import keyring


__version__ = "0.1"
ACCOUNT_MAPPING_FILENAME = os.path.expanduser("~/.aws/accounts")
KEYRING_SERVICE_NAME = "aws_session"


def _boto3_session_closure():
    session = None

    def closure():
        nonlocal session
        if session is None:
            session = boto3.session.Session()
        return session

    return closure


boto3_session = _boto3_session_closure()


def cache_in_keyring(function):
    @wraps(function)
    def wrapper(*args, **kwargs):
        if kwargs.get("account"):
            username = f"{kwargs['account']}_{kwargs['role']}"
        else:
            username = ""

        serialized = keyring.get_password(
            service_name=KEYRING_SERVICE_NAME, username=username
        )
        if serialized:
            token = pickle.loads(base64.b64decode(serialized))
            check_time = datetime.now(timezone.utc) + timedelta(minutes=1)
            if check_time > token["Expiration"]:
                serialized = None
        if serialized is None:
            token = function(*args, **kwargs)
            serialized = base64.b64encode(
                pickle.dumps(token, protocol=pickle.HIGHEST_PROTOCOL)
            ).decode("utf-8")
            keyring.set_password(
                password=serialized,
                service_name=KEYRING_SERVICE_NAME,
                username=username,
            )

        return token

    return wrapper


def clean_environment():
    for name in ("AWS_ACCESS_KEY_ID", "AWS_SECRET_ACCESS_KEY", "AWS_SESSION_TOKEN"):
        if name in os.environ:
            del os.environ[name]


def discover_shell():
    return os.environ.get("SHELL", "/bin/sh")


def handle_cleanup(arguments):
    if arguments.command == "role":
        username = f"{arguments.account}_{arguments.role}"
    else:
        username = ""
    try:
        keyring.delete_password(service_name=KEYRING_SERVICE_NAME, username=username)
    except keyring.errors.PasswordDeleteError:
        pass


@cache_in_keyring
def role_token(client, *, account, role):
    response = client.assume_role(
        DurationSeconds=3600,
        ExternalId=account,
        RoleArn=f"arn:aws:iam::{account}:role/{role}",
        RoleSessionName=os.environ.get("USER", "__"),
    )
    return response["Credentials"]


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--cleanup", action="store_true", help="Delete cached token from keychain."
    )
    parser.add_argument(
        "-s",
        "--shell",
        help="The shell to exec (default: %(default)s).",
        default=discover_shell(),
    )
    parser.add_argument(
        "--version", action="version", version=f"%(prog)s {__version__}"
    )

    subparsers = parser.add_subparsers(dest="command")
    role_parser = subparsers.add_parser("role")
    role_parser.add_argument("role")
    role_parser.add_argument(
        "account",
        help=f"Either an AWS account number, or the name of a key that maps to an account number in {ACCOUNT_MAPPING_FILENAME}.",
        type=validate_account,
    )

    arguments = parser.parse_args()
    if arguments.cleanup:
        handle_cleanup(arguments)
        return 0

    clean_environment()
    token = session_token()
    if arguments.command == "role":
        client = boto3_session().client(
            aws_access_key_id=token["AccessKeyId"],
            aws_secret_access_key=token["SecretAccessKey"],
            aws_session_token=token["SessionToken"],
            service_name="sts",
        )
        token = role_token(client, account=arguments.account, role=arguments.role)
    set_environment(token)

    os.execlp(arguments.shell, arguments.shell)


def mfa_serial_number():
    client = boto3_session().client("iam")
    response = client.list_mfa_devices()
    devices = response["MFADevices"]
    if not devices:
        sys.stderr.write(f"No MFA devices found.\n")
        sys.exit(1)
    assert (
        len(devices) == 1
    ), "How is it possible that there is more than one MFA device?"
    return devices[0]["SerialNumber"]


@cache_in_keyring
def session_token():
    client = boto3_session().client("sts")
    response = None
    while response is None:
        mfa_serial = mfa_serial_number()
        mfa_token = input("MFA Token: ")
        try:
            response = client.get_session_token(
                DurationSeconds=129600, SerialNumber=mfa_serial, TokenCode=mfa_token,
            )
        except (
            botocore.exceptions.ClientError,
            botocore.exceptions.ParamValidationError,
        ) as exception:
            print(exception)
    return response["Credentials"]


def set_environment(token):
    os.environ["AWS_ACCESS_KEY_ID"] = token["AccessKeyId"]
    os.environ["AWS_SECRET_ACCESS_KEY"] = token["SecretAccessKey"]
    os.environ["AWS_SESSION_TOKEN"] = token["SessionToken"]


def validate_account(account):
    if account.isnumeric():
        return account

    try:
        with open(ACCOUNT_MAPPING_FILENAME) as fp:
            accounts = json.load(fp)
    except FileNotFoundError as exception:
        raise argparse.ArgumentTypeError(exception)
    except json.JSONDecodeError:
        raise argparse.ArgumentTypeError(
            f"{fp.name} does not appear to be a valid JSON file"
        )

    if not isinstance(accounts, dict):
        raise argparse.ArgumentTypeError(
            f"{ACCOUNT_MAPPING_FILENAME} is not in the correct format"
        )

    if account not in accounts:
        raise argparse.ArgumentTypeError(
            f"{account} is not found in {ACCOUNT_MAPPING_FILENAME}"
        )

    return accounts[account]


if __name__ == "__main__":
    sys.exit(main())


aws_longer/__init__.py
import argparse
import base64
import json
import os
import pickle
import sys
from datetime import datetime, timedelta, timezone
from functools import wraps

import boto3
import botocore
import keyring


__version__ = "0.4.0"
ACCOUNT_MAPPING_FILENAME = os.path.expanduser("~/.aws/accounts")
KEYRING_SERVICE_NAME = "aws_longer"
ROLE_TOKEN_DURATION = 3600
SESSION_TOKEN_DURATION = 129600
SESSION_KEYRING_USERNAME = "__SESSION__"


def _boto3_session_closure():
    session = None

    def closure():
        nonlocal session
        if session is None:
            session = boto3.session.Session()
        return session

    return closure


boto3_session = _boto3_session_closure()


def cache_in_keyring(function):
    @wraps(function)
    def wrapper(*args, **kwargs):
        if kwargs.get("account"):
            username = f"{kwargs['account']}_{kwargs['role']}"
        else:
            username = SESSION_KEYRING_USERNAME

        serialized = keyring.get_password(
            service_name=KEYRING_SERVICE_NAME, username=username
        )
        if serialized:
            token = pickle.loads(base64.b64decode(serialized))
            check_time = datetime.now(timezone.utc) + timedelta(minutes=1)
            if check_time > token["Expiration"]:
                serialized = None
        if serialized is None:
            token = function(*args, **kwargs)
            serialized = base64.b64encode(
                pickle.dumps(token, protocol=pickle.HIGHEST_PROTOCOL)
            ).decode("utf-8")
            keyring.set_password(
                password=serialized,
                service_name=KEYRING_SERVICE_NAME,
                username=username,
            )

        return token

    return wrapper


def clean_environment():
    for name in ("AWS_ACCESS_KEY_ID", "AWS_SECRET_ACCESS_KEY", "AWS_SESSION_TOKEN"):
        if name in os.environ:
            del os.environ[name]


def discover_shell():
    return os.environ.get("SHELL", "/bin/sh")


def get_aws_username():
    return boto3_session().client("iam").get_user()["User"]["UserName"]


def get_token(arguments, mfa_token_callback):
    if arguments.command == "role":

        def client_callback():
            token = session_token(mfa_token_callback=mfa_token_callback)
            return boto3_session().client(
                aws_access_key_id=token["AccessKeyId"],
                aws_secret_access_key=token["SecretAccessKey"],
                aws_session_token=token["SessionToken"],
                service_name="sts",
            )

        token = role_token(
            client_callback, account=arguments.account, role=arguments.role
        )
    else:
        token = session_token(mfa_token_callback=mfa_token_callback)
    return token


def handle_cleanup(arguments):
    if arguments.command == "role":
        username = f"{arguments.account}_{arguments.role}"
    else:
        username = SESSION_KEYRING_USERNAME
    try:
        keyring.delete_password(service_name=KEYRING_SERVICE_NAME, username=username)
    except keyring.errors.PasswordDeleteError:
        pass


@cache_in_keyring
def role_token(client_callback, *, account, role):
    response = client_callback().assume_role(
        DurationSeconds=ROLE_TOKEN_DURATION,
        ExternalId=account,
        RoleArn=f"arn:aws:iam::{account}:role/{role}",
        RoleSessionName=get_aws_username(),
    )
    return response["Credentials"]


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--cleanup", action="store_true", help="Delete cached token from keychain."
    )
    parser.add_argument(
        "-s",
        "--shell",
        help="The shell to exec (default: %(default)s).",
        default=discover_shell(),
    )
    parser.add_argument(
        "-m",
        "--mfa-token",
        help="When an MFA token is required, pass in this value instead of prompting for it.",
    )
    parser.add_argument(
        "--version", action="version", version=f"%(prog)s {__version__}"
    )
    parser.add_argument(
        "--yubikey",
        help="When an MFA token is required, obtain it from the attached yubikey using the provided name.",
    )

    subparsers = parser.add_subparsers(dest="command")
    role_parser = subparsers.add_parser("role")
    role_parser.add_argument("role")
    role_parser.add_argument(
        "account",
        help=f"Either an AWS account number, or the name of a key that maps to an account number in {ACCOUNT_MAPPING_FILENAME}.",
        type=validate_account,
    )

    arguments = parser.parse_args()
    if arguments.mfa_token and arguments.yubikey:
        parser.error("Cannot provide both --mfa-token and --yubikey")
        return 1

    if arguments.cleanup:
        handle_cleanup(arguments)
        return 0

    def mfa_token_callback():
        if arguments.mfa_token:
            return arguments.mfa_token
        if arguments.yubikey is None:
            return None
        try:
            import ykman  # noqa: F401
        except ModuleNotFoundError:
            print(
                "Please ensure ykman is installed. Try: `pip install aws_longer[yubikey]`"
            )
            return None
        import subprocess

        process = subprocess.run(
            ["ykman", "oath", "code", "--single", arguments.yubikey],
            stdout=subprocess.PIPE,
        )
        if process.returncode != 0:
            raise GracefulExit
        return process.stdout.strip().decode("utf-8")

    clean_environment()

    try:
        token = get_token(arguments, mfa_token_callback)
    except GracefulExit:
        print("Leaving aws_longer.")
        token = None

    if token:
        set_environment(token)

    os.execlp(arguments.shell, arguments.shell)


def mfa_serial_number():
    client = boto3_session().client("iam")
    response = client.list_mfa_devices()
    devices = response["MFADevices"]
    if not devices:
        sys.stderr.write("No MFA devices found.\n")
        sys.exit(1)
    assert (
        len(devices) == 1
    ), "How is it possible that there are more than one MFA device?"
    return devices[0]["SerialNumber"]


@cache_in_keyring
def session_token(mfa_token_callback):
    client = boto3_session().client("sts")
    mfa_serial = mfa_serial_number()
    mfa_token = mfa_token_callback()
    response = None
    while response is None:
        try:
            mfa_token = mfa_token or input("MFA Token: ")
        except EOFError:
            raise GracefulExit
        except KeyboardInterrupt:
            sys.exit(1)
        try:
            response = client.get_session_token(
                DurationSeconds=SESSION_TOKEN_DURATION,
                SerialNumber=mfa_serial,
                TokenCode=mfa_token,
            )
        except (
            botocore.exceptions.ClientError,
            botocore.exceptions.ParamValidationError,
        ) as exception:
            print(exception)
            mfa_token = None
    return response["Credentials"]


def set_environment(token):
    os.environ["AWS_ACCESS_KEY_ID"] = token["AccessKeyId"]
    os.environ["AWS_SECRET_ACCESS_KEY"] = token["SecretAccessKey"]
    os.environ["AWS_SESSION_TOKEN"] = token["SessionToken"]


def validate_account(account):
    if account.isnumeric():
        return account

    try:
        with open(ACCOUNT_MAPPING_FILENAME) as fp:
            accounts = json.load(fp)
    except FileNotFoundError as exception:
        raise argparse.ArgumentTypeError(exception)
    except json.JSONDecodeError:
        raise argparse.ArgumentTypeError(
            f"{fp.name} does not appear to be a valid JSON file"
        )

    if not isinstance(accounts, dict):
        raise argparse.ArgumentTypeError(
            f"{ACCOUNT_MAPPING_FILENAME} is not in the correct format"
        )

    if account not in accounts:
        raise argparse.ArgumentTypeError(
            f"{account} is not found in {ACCOUNT_MAPPING_FILENAME}"
        )

    return accounts[account]


class GracefulExit(Exception):
    pass
